### PR TITLE
Support for composite primary keys

### PR DESCRIPTION
**Pull request by sharadregoti**
1) Added support for composite primary keys

### CODE REVIEW

File: gateway/modules/schema/helpers.go [hunk 1]
```diff
@@ -344,6 +346,28 @@ func (s *Schema) addNewTable(ctx context.Context, logicalDBName, dbType, dbAlias
 	if !doesPrimaryKeyExists {
 		return "", helpers.Logger.LogError(helpers.GetRequestID(ctx), "Primary key not found, make sure there is a primary key on a field with type (ID)", nil, nil)
 	}
+
+	finalPrimaryKeyQuery := "PRIMARY KEY ("
```
Comment (YourTechBud) [inline gateway/modules/schema/helpers.go:350]: Put all of this in a helper function

File: gateway/modules/schema/inspection.go [hunk 2]
```diff
@@ -86,17 +86,17 @@ func generateInspection(dbType, col string, fields []model.InspectorFieldType, f
...
-		// Set auto increment
-		if field.AutoIncrement == "true" {
-			fieldDetails.IsAutoIncrement = true
-		}
-		if model.DBType(dbType) == model.Postgres && strings.HasPrefix(field.FieldDefault, "nextval") {
-			// override the default value, this is a special case if a postgres column has a auto increment value, the default value that database returns is -> ( nextval(auto_increment_test_auto_increment_test_seq )
-			fieldDetails.Default = ""
-			fieldDetails.IsDefault = false
-			fieldDetails.IsAutoIncrement = true
+			fieldDetails.PrimaryKeyInfo = &model.TableProperties{}
```
Comment (YourTechBud) [inline gateway/modules/schema/inspection.go:89]: We aren't setting the order here?

File: gateway/modules/schema/schema.go [hunk 3]
```diff
@@ -153,15 +153,28 @@ func getCollectionSchema(doc *ast.Document, dbName, collectionName string) (mode
...
-									fieldTypeStuct.IsAutoIncrement = t
+									fieldTypeStuct.PrimaryKeyInfo.IsAutoIncrement = t
 								default:
 									return nil, helpers.Logger.LogError(helpers.GetRequestID(context.TODO()), fmt.Sprintf("Unexpected argument type provided for field (%s) directinve @(%s) argument (%s) got (%v) expected string", fieldTypeStuct.FieldName, directive.Name.Value, argument.Name.Value, reflect.TypeOf(val)), nil, map[string]interface{}{"arg": argument.Name.Value})
 								}
+
+							}
+							if argument.Name.Value == "order" {
+								val, _ := utils.ParseGraphqlValue(argument.Value, nil)
+								t, ok := val.(int)
```
Comment (YourTechBud) [inline gateway/modules/schema/schema.go:170]: I thought parseGraphqlValue returns int64

File: gateway/modules/schema/template.go [hunk 4]
```diff
@@ -9,7 +9,7 @@ import (
...
...Order}} ) {{else if $fieldValue.PrimaryKeyInfo.IsAutoIncrement}}  (autoIncrement: {{$fieldValue.PrimaryKeyInfo.IsAutoIncrement}}) {{else if $fieldValue.PrimaryKeyInfo.Order}}  (order: {{$fieldValue.PrimaryKeyInfo.Order}}) {{end}}  {{end}}  {{if eq $fieldValue.Kind "ID"}}@size(value: {{$fieldValue.TypeIDSize}}){{end}} {{if $fieldValue.IsCreatedAt}}@createdAt{{end}} {{if $fieldValue.IsUpdatedAt}}@updatedAt{{end}} {{if $fieldValue.IsUnique}}@unique(group: "{{$fieldValue.IndexInfo.Group}}", order: {{$fieldValue.IndexInfo.Order}})  {{else}} {{if $fieldValue.IsIndex}}@index(group: "{{$fieldValue.IndexInfo.Group}}", sort: "{{$fieldValue.IndexInfo.Sort}}", order: {{$fieldValue.IndexInfo.Order}}){{end}}{{end}} {{if $fieldValue.IsDefault}}@default(value: {{$fieldValue.Default}}){{end}} {{if $fieldValue.IsLinked}}@link(table: {{$fieldValue.LinkedTable.Table}}, from: {{$fieldValue.LinkedTable.From}}, to: {{$fieldValue.LinkedTable.To}}, field: {{$fieldValue.LinkedTable.Field}}){{end}} {{if $fieldValue.IsForeign}}@foreign(table: {{$fieldValue.JointTable.Table}}, field: {{$fieldValue.JointTable.To}}{{if eq $fieldValue.JointTable.OnDelete "CASCADE"}} ,onDelete: cascade{{end}}){{end}}{{end}}{{end}}
```
Comment (YourTechBud) [inline gateway/modules/schema/template.go:12]: I bet you can find a way to make this template simpler

File: gateway/modules/schema/helpers.go [hunk 5]
```diff
@@ -292,6 +292,7 @@ func (s *Schema) addNewTable(ctx context.Context, logicalDBName, dbType, dbAlias
 
 	var query, primaryKeyQuery string
 	doesPrimaryKeyExists := false
+	compositePrimaryKeys := make([]*model.FieldType, 0)
```
Comment (YourTechBud) [inline gateway/modules/schema/helpers.go:295]: Make this primary key store

File: gateway/modules/schema/helpers.go [hunk 6]
```diff
@@ -344,6 +346,28 @@ func (s *Schema) addNewTable(ctx context.Context, logicalDBName, dbType, dbAlias
 	if !doesPrimaryKeyExists {
 		return "", helpers.Logger.LogError(helpers.GetRequestID(ctx), "Primary key not found, make sure there is a primary key on a field with type (ID)", nil, nil)
 	}
+
+	finalPrimaryKeyQuery := "PRIMARY KEY ("
+	if len(compositePrimaryKeys) > 1 {
+		var v primaryKeyStore = compositePrimaryKeys
```
Comment (YourTechBud) [inline gateway/modules/schema/helpers.go:352]: Shouldn't be required if `conpositePrimaryKeys` is of type `primaryKeyStore`.

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: gateway/modules/schema/inspection.go (hunk 2 above)
Comment (sharadregoti) [reply gateway/modules/schema/inspection.go:89]: We will do that in other pull requests, I have noted this down

File: gateway/modules/schema/schema.go (hunk 3 above)
Comment (sharadregoti) [reply gateway/modules/schema/schema.go:170]: It is int only
image

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**